launcher/ui/dialogs/InstallLoaderDialog.cpp

```diff
@@ -0,0 +1,164 @@
+
+    void openedImpl() override
+    {
+        if (loaded)
+            return;
+
+        const auto versions = APPLICATION->metadataIndex()->get(uid);
+        if (!versions)
+            return;
+
+        initialize(versions.get());
+        loaded = true;
+    }
+
+    void setParentContainer(BasePageContainer* container) override
+    {
+        auto dialog = dynamic_cast<QDialog*>(dynamic_cast<PageContainer*>(container)->parent());
+        connect(view(), &QAbstractItemView::doubleClicked, dialog, &QDialog::accept);
+    }
+
+   private:
+    const QString uid;
+    const QString iconName;
+    const QString name;
+    bool loaded = false;
+};
+
+static InstallLoaderPage* pageCast(BasePage* page)
+{
+    auto result = dynamic_cast<InstallLoaderPage*>(page);
+    Q_ASSERT(result != nullptr);
+    return result;
+}
+
+InstallLoaderDialog::InstallLoaderDialog(std::shared_ptr<PackProfile> profile, const QString& uid, QWidget* parent)
+    : QDialog(parent), profile(std::move(profile)), container(new PageContainer(this, QString(), this)), buttons(new QDialogButtonBox(this))
+{
+    auto layout = new QVBoxLayout(this);
+
+    container->setSizePolicy(QSizePolicy::Preferred, QSizePolicy::Expanding);
+    layout->addWidget(container);
+
+    auto buttonLayout = new QHBoxLayout(this);
+
+    auto refreshButton = new QPushButton(tr("&Refresh"), this);
+    connect(refreshButton, &QPushButton::pressed, this, [this] { pageCast(container->selectedPage())->loadList(); });
+    buttonLayout->addWidget(refreshButton);
+
+    buttons->setOrientation(Qt::Horizontal);
+    buttons->setStandardButtons(QDialogButtonBox::Cancel | QDialogButtonBox::Ok);
+    connect(buttons, &QDialogButtonBox::accepted, this, &QDialog::accept);
+    connect(buttons, &QDialogButtonBox::rejected, this, &QDialog::reject);
+    buttonLayout->addWidget(buttons);
+
+    layout->addLayout(buttonLayout);
+
+    setWindowTitle(dialogTitle());
+    setWindowModality(Qt::WindowModal);
+    resize(520, 347);
+
+    for (BasePage* page : container->getPages()) {
+        if (page->id() == uid)
+            container->selectPage(page->id());
+
+        connect(pageCast(page), &VersionSelectWidget::selectedVersionChanged, this, [this, page] {
+            if (page->id() == container->selectedPage()->id())
+                validate(container->selectedPage());
+        });
+    }
+    connect(container, &PageContainer::selectedPageChanged, this, [this](BasePage* previous, BasePage* current) { validate(current); });
+    pageCast(container->selectedPage())->selectSearch();
+    validate(container->selectedPage());
+}
+
+QList<BasePage*> InstallLoaderDialog::getPages()
+{
+    return { // Forge
+             new InstallLoaderPage("net.minecraftforge", "forge", tr("Forge"), {}, profile),
+             // Fabric
+             new InstallLoaderPage("net.fabricmc.fabric-loader", "fabricmc-small", tr("Fabric"), Version("1.14"), profile),
```

Comment:
why -small? :O

Reply:
smol

Reply:
:iea: It's because i needed multiple sizes because scaling pixel art is hard. and adding 22x22/instances might not have worked... i can't remember.